File: problems/0027_Remove_Element/removeElements.cpp
/*

 题目：删除链表中给定值val的所有节点
 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        
      
          while(head != NULL )
        {
            if(head->val == val)
            {
                head = head->next;
            }
            else
                break;
        }
        
        ListNode*s = head;
        ListNode*p = head;
        while(s != NULL)
        {
            if(s->val == val)
            {
                p->next = s->next;
                s = p->next;
                continue;
            }
            p = s;
            s = s->next;
        }
        return head;
    }
};

File: problems/0203_Remove_Linked_List_Elements/RemoveElements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        if head is None:
            return head
        while head is not None and head.val == val:
            head = head.next;
        list_head = head
        pre = head
        while head is not None:
            if head.val == val:
                pre.next = head.next
                head = head.next
            else:
                pre = head
                head = head.next
        return list_head
        
